packages/@react-stately/numberfield/src/useNumberFieldState.ts

```diff
@@ -105,9 +104,17 @@ export function useNumberFieldState(
   // Update the input value when the number value or format options change. This is done
   // in a useEffect so that the controlled behavior is correct and we only update the
   // textfield after prop changes.
-  useLayoutEffect(() => {
+  let prevValue = useRef(numberValue);
+  let prevLocale = useRef(locale);
+  let prevFormatOptions = useRef(formatOptions);
+  // because NaN === NaN => false
+  let notEquivalent = isNaN(prevValue.current) || isNaN(numberValue) ? Boolean(isNaN(prevValue.current)) !== Boolean(isNaN(numberValue)) : numberValue !== prevValue.current;
```

Comment:
You can use `Object.is` for this: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is